Added seeds to `evaluator.simple_evaluate` signature

Am1n3e opened the pull request:
With the latest changes (e.g., `TaskManager`) it is much easier to integrate `lm_eval` into existing code.
However, the `evaluator.simple_evalute` function sets `random`, `numpy`, and `torch` seeds with hard-coded values.

Basically this PR addresses this TODO item:

https://github.com/EleutherAI/lm-evaluation-harness/blob/4c17c55c9ae3b22280daf4046f70c176e62706d4/lm_eval/evaluator.py#L91

Changes in this PR:

- Make seed values as parameters so the user can choose the values, or set it to `None` if setting the seed is done outside of `lm-eval` when it is used as a library.
- Keep the same values as default values for backward compatibility.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1412) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/EleutherAI/lm-evaluation-harness?pullRequest=1412) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/EleutherAI/lm-evaluation-harness?pullRequest=1412) it.</sub>

christyler3030 commented:
Works for me and doesn't change the default behaviour.

Am1n3e commented:
@haileyschoelkopf thanks for the feedback.

Regarding the cli arg, that was one reason why I did not include it initially (to avoid bloating) the CLI interface. 

Here what I propose: instead of adding an argument we can have a '--simple_eval_extra_args'. This is the same spirit of '--model_args'.
The extra args will be converted to kwargs like for models. Python will throw an error if the user tries yo set a value that set by '__main__'.

Example:

'lm_eval ... --simple_eval_extra_args numpy_seed=1,torch_seed=null'

In the example above the user wants to keep the default python random seed to the default value.

This way we can solve passing other args without any changes to the CLI args.

Let me know what you think. I don't mind adding the seed arg you proposed (single int, single'None' or list mixed list

haileyschoelkopf commented:
@Am1n3e I'll think on the `--simple_eval_extra_args` proposal!

I think for now I'd prefer if we can stick with just the single `seed` kwarg and CLI arg for simplicity's sake, but may go back and allow for this arbitrary-argument feature later if it seems useful.

Am1n3e commented on lm_eval/evaluator.py:
```diff
@@ -80,18 +83,31 @@ def simple_evaluate(
 
+    if random_seed is not None:
+        # See https://github.com/EleutherAI/lm-evaluation-harness/pull/1412
+        eval_logger.info(f"Setting random seed to {random_seed}")
```
Added logs when the seeds are set to notify the users that this happening since it took me a while to figure this out.

Am1n3e replied:
Just an FYI.

haileyschoelkopf replied:
Thank you for doing so!

Am1n3e commented:
@haileyschoelkopf  I have added the CLI argument as we discussed. Below is the `-h` output for the newly added arg for reference.  Further, I have added logs when the seed is set to alert users that simple evaluate sets the seed.

```
 --seed SEED           Set seed for python's random, numpy and torch.
                        Accepts a comma-separated list of 3 values for python's random, numpy, and torch seeds, respectively, or a single integer to set the same seed for all three.
                        The values are either an integer or 'None' to not set the seed. Default is `0,1234,1234` (for backward compatibility).
                        E.g. `--seed 0,None,8` sets `random.seed(0)` and `torch.manual_seed(8)`. Here numpy's seed is not set since the second value is `None`.
                        E.g, `--seed 42` sets all three seeds to 42.
```

Am1n3e commented:
@haileyschoelkopf  Regarding setting additional args discussion. First, what I proposed was just a quick hack. I think the library would benefit from proper handling for setting parameters more easily. Because, if you look at some of the forks, users sometimes want to set specific parameters that are not available in the CLI and end up writing custom entry points or mocking some functions. While that solves the user problem it can make updating to a new `lm_eval` version a hassle. Also, once forked, `lm_eval` loses the possibility of users submitting PRs since it is much easier to push in their own fork.

I think configuration management like [hydra](https://hydra.cc/) would make the lib much more flexible while doing close to no effort on boilerplate code for config management using a YAML file or from the CLI. Of course, there are some changes that need to be made. But I think an initial version (that would solve the extra args, for example) would be easy to do.

I will be happy to discuss this further and see how we can collaborate on this.

haileyschoelkopf commented:
@Am1n3e I like the idea of an alternate entry point! And agree with your sentiment that we want it to be easy to upstream changes or new tasks contributed by users.

If you have discord, let's discuss further on [EleutherAI's discord](discord.gg/eleutherai) in the #lm-thunderdome channel! Does that work for you?

Would you also be willing to add these seed args to `docs/interface.md` before we merge? If not, happy to do so myself.

Am1n3e commented:
@haileyschoelkopf I've updated `interface.md`

haileyschoelkopf commented:
Thank you very much!

LSinev commented:
May we also have control over fewshot context sampler seed 
here https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/api/task.py#L523 (for Task)
and here https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/api/task.py#L815 (for ConfigurableTask)?
